pkg/protocol/gossip/service.go

```diff
@@ -0,0 +1,447 @@
+			count++
+		}
+		return true
+	})
+	return count
+}
+
+// Start starts the Service's event loop.
+func (s *Service) Start(shutdownSignal <-chan struct{}) {
+	s.host.SetStreamHandler(s.protocol, func(stream network.Stream) {
+		s.inboundStreamChan <- stream
+	})
+	s.manager.Events.Connected.Attach(events.NewClosure(func(p *p2p.Peer, conn network.Conn) {
+		s.connectedChan <- &connectionmsg{p: p, conn: conn}
+	}))
+	s.manager.Events.Disconnected.Attach(events.NewClosure(func(p *p2p.Peer) {
+		s.disconnectedChan <- &connectionmsg{p: p, conn: nil}
+	}))
+	s.manager.Events.RelationUpdated.Attach(events.NewClosure(func(p *p2p.Peer, oldRel p2p.PeerRelation) {
+		s.relationUpdatedChan <- &relationupdatedmsg{peer: p, oldRelation: oldRel}
+	}))
+	// manage libp2p network events
+	s.host.Network().Notify((*netNotifiee)(s))
+	s.eventLoop(shutdownSignal)
+	// de-register libp2p network events
+	s.host.Network().StopNotify((*netNotifiee)(s))
+}
+
+type connectionmsg struct {
+	p    *p2p.Peer
+	conn network.Conn
+}
+
+type streamreqmsg struct {
+	p    peer.ID
+	back chan *Protocol
+}
+
+type streamclosedmsg struct {
+	p      peer.ID
+	stream network.Stream
+}
+
+type relationupdatedmsg struct {
+	peer        *p2p.Peer
+	oldRelation p2p.PeerRelation
+}
+
+type foreachmsg struct {
+	f    ProtocolForEachFunc
+	back chan struct{}
+}
+
+// runs the Service's event loop, handling inbound/outbound streams.
+func (s *Service) eventLoop(shutdownSignal <-chan struct{}) {
+	for {
+		select {
+		case <-shutdownSignal:
+			return
+
+		case inboundStream := <-s.inboundStreamChan:
+			if proto := s.handleInboundStream(inboundStream); proto != nil {
+				s.Events.ProtocolStarted.Trigger(proto)
+			}
+
+		case connectedMsg := <-s.connectedChan:
+			proto, err := s.handleConnected(connectedMsg.p, connectedMsg.conn)
+			if err != nil {
+				s.Events.Error.Trigger(err)
+				continue
+			}
+
+			if proto != nil {
+				s.Events.ProtocolStarted.Trigger(proto)
+			}
+
+		case disconnectedMsg := <-s.disconnectedChan:
+			s.closeStream(disconnectedMsg.p.ID)
+
+		case streamClosedMsg := <-s.streamClosedChan:
+			s.closeStream(streamClosedMsg.p)
+
+		case relationUpdatedMsg := <-s.relationUpdatedChan:
+			proto, err := s.handleRelationUpdated(relationUpdatedMsg.peer, relationUpdatedMsg.oldRelation)
+			if err != nil {
+				s.Events.Error.Trigger(err)
+				continue
+			}
+
+			if proto != nil {
+				s.Events.ProtocolStarted.Trigger(proto)
+			}
+
+		case streamReqMsg := <-s.streamReqChan:
+			streamReqMsg.back <- s.proto(streamReqMsg.p)
+
+		case forEachMsg := <-s.forEachChan:
+			s.forEach(forEachMsg.f)
+			forEachMsg.back <- struct{}{}
+		}
+	}
+}
+
+// handles incoming streams and closes them if the given peer's relation should not allow any.
+func (s *Service) handleInboundStream(stream network.Stream) *Protocol {
+	remotePeerID := stream.Conn().RemotePeer()
+	// close if there is already one
+	if _, ongoing := s.streams[remotePeerID]; ongoing {
+		s.closeUnwantedStream(stream)
```

Comment:
Should the 
`s.Events.InboundStreamCancelled.Trigger(stream)`
be called here as well?

Reply:
Sure, I've also added a reason for the cancellation, so there is a distinction between the two triggers.